Make username case-insensitive on login

KroniK907 opened the pull request:
Both the username submitted by the user, and the username associated with each user has casefold applied to make the username case-insensitive at login. Using casefold instead of lower to ensure that it works for edge cases in other languages as well as english. 

Casefold was also applied when checking if an existing username already exists at user creation, as well as when doing a password update, as well as when removing a user. This ensures that in all situations, there can only be a single user with a given case-insensitive username.

99% of web logins make the username case-insensitive, and having the login for hassio be case-sensitive can be both confusing and frustrating for admins, and users alike. As case-insensitive usernames are expected by most users, it makes sense to make the login process conform to most users expectations.

KroniK907 committed:
Make username case-insensitive on login

Both the username submitted by the user, and the username associated with each user has casefold applied to make the username case-insensitive at login. Using casefold instead of lower to ensure that it works for edge cases in other languages as well as english.

Casefold was also applied when checking if an existing username already exists at user creation, as well as when doing a password update, as well as when removing a user. This ensures that in all situations, there can only be a single user with a given case-insensitive username.

99% of web logins make the username case-insensitive, and having the login for hassio be case-sensitive can be both confusing and frustrating for admins, and users alike. As case-insensitive usernames are expected by most users, it makes sense to make the login process conform to most users expectations.

homeassistant commented:
Hi @KroniK907, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#19237).

Once you do that we will be able to review and accept this pull request.

Thanks!

KroniK907 commented:
Because usernames were case sensitive at rollout of the auth system, it is possible that some people have 2 or more users that have usernames which would be in conflict with a case-insensitive system. However, I think conforming to standard web login practices, and making the login system meet most users expectations is of more importance.

KroniK907 commented:
force pushed a quick formatting edit because travis doesn't like lines over 79 characters

KroniK907 committed:
Add testing for case-insensitive usernames

Added 2 test cases. First is to verify that a new user cannot be created
if the chosen username already exists but with different capitalization.
Second is to verify that a user can login by capitalizing letters in
their username.

balloob commented:
I agree that this is a good change. However we need to include a migration path as this can get people locked out of their system. We need to migrate duplicate usernames to be appended with a 2/3/4 etc.

KroniK907 commented:
Definitely is a bit of a bummer. I don't have the skills to add a whole migration system.

Maybe it can just be added as an admin option with lots of warnings?

balloob commented:
We're not going to make it an option, it should just be fixed. 

The homeassistant provider uses the storage helper which has support for migrations.

Add to `homeassistant.py`:

```python
from homeassistant.helpers.storage import Store

STORAGE_VERSION = 2

class HomeAssistantProviderStore(Store):

    async def _async_migrate_func(self, old_version, old_data):
        # Migration logic here
```

Then replace the `self._store` assignment:

```python
        self._store = HomeAssistantProviderStore(STORAGE_VERSION, STORAGE_KEY, private=True)
```

KroniK907 commented:
So I am but a humble sysadmin with just enough programming knowledge to be dangerous. Can you point me to either documentation or the location in the code where I could study the old user store format and how the data is formatted/stored? 

I think it would just take a couple for loops to do the migration, but I'm not sure of the current structure. Also, How do I ensure that this migration only happens once? Is there an existing storage version that is set that I can check before running the migration? Or is that built into the migration function you referenced?

frenck commented:
Just an IMHO: A username should be case sensitive. Frenck != frenck.

> 99% of web logins make the username case-insensitive

Could you please refer me to reference material where you base this on?

Small extra note: This adds to lowering security, as it drops the attack surface as well.

KroniK907 commented:
@frenck 
All major sites that people are familiar with eg: Google, facebook, amazon, spotify, adobe, github, etc, all use case insensitive usernames or email addresses as the unique user identifier. This was literally the first web based app that I have ever used (besides a limited number of LDAP based logins) that used a case sensitive username.

TheBestMoshe commented:
It's especially confusing to have case sensitive usernames, as when logging in on mobile, generally, the first character is auto-capitalized. Which has the potential to lead to lots of confusion.

balloob commented:
I cannot think of a valid usecase where having two users with the same username but different capitalization is useful. I'm fine with restricting it, if a migration is included with the PR. 

Real life example is Google accounts. They actually go a step further, the following emails all end up in my inbox:
 - balloob@gmail.com
 - Balloob@gmail.com
 - ball.o.o.b@gmail.com
 - b.al.lo.ob+otherstuff@gmail.com

balloob commented:
I will finish this PR.

balloob commented:
I've opened PR #20150 to add a foundation to have the auth be able to run in two different modes. One modern mode that strips and a legacy mode that doesn't strip whitespace. The legacy mode is automatically activated if a username with a space is detected. Once #20150 is merged, this PR will need to be rebased and add casefolding to the modern mode.

balloob commented:
Going to close this. Let's continue discussion first in architecture repo https://github.com/home-assistant/architecture/issues/134
